src/table/sparse_matrix_table.cpp

```diff
@@ -9,290 +10,299 @@
+  }
 
-    int row_size = sizeof(T)* num_col_;
-    result->push_back(Blob(keys_size * row_size));
-    T* vals = reinterpret_cast<T*>((*result)[1].data());
-    int offset_v = 0;
-    for (int i = 0; i < keys_size; ++i) {
-      int offset_s = (keys[i] - row_offset_) * num_col_;
-      memcpy(vals + offset_v, &storage_[offset_s], row_size);
-      offset_v += num_col_;
-      Log::Debug("[ProcessAdd] Server = %d, getting #row = %d\n",
-        server_id_, keys[i]);
+  if (key_size == 1 && keys[0] == -1) {
+    for (auto local_row_id = 0; local_row_id < my_num_row_; ++local_row_id)  {
+      if (!up_to_date_[worker_id][local_row_id]) {
+        out_rows->push_back(get_global_row_id(local_row_id));
+        up_to_date_[worker_id][local_row_id] = true;
+      }
+    }
+  }
+  else {
+    for (auto i = 0; i < key_size; ++i)  {
+      auto global_row_id = keys[i];
+      auto local_row_id = get_local_row_id(global_row_id);
+      if (!up_to_date_[worker_id][local_row_id]) {
+        up_to_date_[worker_id][local_row_id] = true;
+        out_rows->push_back(global_row_id);
+      }
     }
   }
 
-  template <typename T>
-  void SparseMatrixServerTable<T>::Store(Stream* s) {
-    s->Write(storage_.data(), storage_.size() * sizeof(T));
+  // if all rows are up-to-date, then send the first row
+  if (out_rows->size() == 0) {
+    out_rows->push_back(get_global_row_id(0));
   }
+}
 
-  template <typename T>
-  void SparseMatrixServerTable<T>::Load(Stream* s) {
-    s->Read(storage_.data(), storage_.size() * sizeof(T));
+template <typename T>
+void SparseMatrixServerTable<T>::ProcessAdd(
+  const std::vector<Blob>& compressed_data) {
+  std::vector<Blob> data;
+  SparseFilter<T, int32_t> filter(0, true);
+  filter.FilterOut(compressed_data, &data);
+
+  // the UpdateOption option is needed for the sparse update
+  CHECK(data.size() == 3);
+  UpdateOption* option = nullptr;
+  if (data.size() == 3) {
+    option = new UpdateOption(data[2].data(), data[2].size());
   }
+  UpdateAddState(option->worker_id(), data[0]);
+  MatrixServerTable<T>::ProcessAdd(data);
+}
 
-  MV_INSTANTIATE_CLASS_WITH_BASE_TYPE(SparseMatrixWorkerTable);
-  MV_INSTANTIATE_CLASS_WITH_BASE_TYPE(SparseMatrixServerTable);
+template <typename T>
+void SparseMatrixServerTable<T>::ProcessGet(
+  const std::vector<Blob>& compressed_data,
+  std::vector<Blob>* result) {
+  std::vector<Blob> data;
+  SparseFilter<T, int32_t> filter(0, true);
+  filter.FilterOut(compressed_data, &data);
+
+  // the general option is needed for the sparse update
+  CHECK(data.size() == 2);
+  CHECK_NOTNULL(result);
+
+  size_t keys_size = data[0].size<int>();
+  int *keys = reinterpret_cast<int*>(data[0].data());
+  GeneralOption* option = nullptr;
+  if (data.size() == 2) {
+    option = new GeneralOption(data[1].data(), data[1].size());
+  }
+  std::vector<int> outdate_rows;
+#pragma warning( push )
+#pragma warning( disable : 4267)
+  UpdateGetState(option->worker_id(), keys, keys_size, &outdate_rows);
```

Comment:
remove disable warning. Here the code assume the option is not null, which means data.size always equal to 2? In this case maybe you can use CHECK and remove the if condition.